cmd/launcher/watchdog/controller_windows.go

```diff
@@ -0,0 +1,328 @@
+		wc.slogger.LogAttrs(ctx, slog.LevelInfo, "", logArgs...)
+		logsToDelete = append(logsToDelete, rowid)
+
+		return nil
+	}); err != nil {
+		wc.slogger.Log(ctx, slog.LevelError, "iterating sqlite logs", "err", err)
+		return
+	}
+
+	if len(logsToDelete) == 0 { // nothing else to do if there are no new logs
+		return
+	}
+
+	wc.slogger.Log(ctx, slog.LevelInfo, "collected logs for deletion", "rowids", logsToDelete)
+
+	if err := wc.logPublisher.DeleteRows(logsToDelete...); err != nil {
+		wc.slogger.Log(ctx, slog.LevelError, "cleaning up published sqlite logs", "err", err)
+	}
+}
+
+func (wc *WatchdogController) Interrupt(_ error) {
+	// Only perform shutdown tasks on first call to interrupt -- no need to repeat on potential extra calls.
+	if wc.interrupted {
+		return
+	}
+
+	wc.logPublisher.Close()
+	wc.interrupted = true
+	wc.interrupt <- struct{}{}
+}
+
+func (wc *WatchdogController) ServiceEnabledChanged(enabled bool) {
+	ctx := context.TODO()
+
+	serviceManager, err := mgr.Connect()
+	if err != nil {
+		wc.slogger.Log(ctx, slog.LevelError,
+			"connecting to service control manager",
+			"err", err,
+		)
+
+		return
+	}
+
+	defer serviceManager.Disconnect()
+
+	if !enabled {
+		err := removeService(serviceManager, launcherWatchdogServiceName)
+		if err != nil && err.Error() != serviceDoesNotExistError {
+			wc.slogger.Log(ctx, slog.LevelWarn,
+				"encountered error removing watchdog service",
+				"err", err,
+			)
+		}
+
+		wc.slogger.Log(ctx, slog.LevelInfo, "removed watchdog service")
+
+		return
+	}
+
+	// we're enabling the watchdog - first check if we've already installed the service
+	// there are three potential paths here:
+	// 1. service did not previously exist, proceed with clean installation
+	existingService, err := serviceManager.OpenService(launcherWatchdogServiceName)
+	if err != nil && err.Error() == serviceDoesNotExistError {
+		if err = wc.installService(serviceManager); err != nil {
+			wc.slogger.Log(ctx, slog.LevelError,
+				"encountered error installing watchdog service",
+				"err", err,
+			)
+		}
+
+		return
+	}
+
+	// 2. we are unable to check the current status of the service,
+	// this is the least likely option and there's nothing we can do here so log and return
+	if err != nil {
+		wc.slogger.Log(ctx, slog.LevelWarn,
+			"encountered error checking for watchdog service, unable to proceed with enabling",
+			"err", err,
+		)
+
+		return
+	}
+
+	// 3. The watchdog service already exists on this device. Here we just restart it to ensure it is
+	// running on the latest launcher code
+	defer existingService.Close()
+	if err = wc.restartService(existingService); err != nil {
+		wc.slogger.Log(ctx, slog.LevelError,
+			"failure attempting to restart watchdog service",
+			"err", err,
+		)
+	}
+}
+
+func (wc *WatchdogController) installService(serviceManager *mgr.Mgr) error {
+	ctx := context.TODO()
+	currentExe, err := os.Executable()
+	if err != nil {
```

Comment:
should we return here if we don't get the executable?

Reply:
yes, thank you!